src/index/CompressedBlockPrefiltering.cpp

```diff
@@ -0,0 +1,278 @@
+//  Copyright 2024, University of Freiburg,
+//                  Chair of Algorithms and Data Structures
+//  Author: Hannes Baumann <baumannh@informatik.uni-freiburg.de>
+
+#include "index/CompressedBlockPrefiltering.h"
+
+#include "global/ValueIdComparators.h"
+
+namespace prefilterExpressions {
+
+// HELPER FUNCTIONS
+//______________________________________________________________________________
+// Given a PermutedTriple retrieve the suitable Id w.r.t. a column (index).
+static Id getIdFromColumnIndex(const BlockMetadata::PermutedTriple& triple,
+                               size_t columnIndex) {
+  switch (columnIndex) {
+    case 0:
+      return triple.col0Id_;
+    case 1:
+      return triple.col1Id_;
+    case 2:
+      return triple.col2Id_;
+    default:
+      // columnIndex out of bounds
+      AD_FAIL();
+  }
+};
+
+//______________________________________________________________________________
+// Extract the Ids from the given `PermutedTriple` in a tuple w.r.t. the
+// position (column index) defined by `ignoreIndex`. The ignored positions are
+// filled with Ids `Id::min()`. `Id::min()` is guaranteed
+// to be smaller than Ids of all other types.
+static auto getMaskedTriple(const BlockMetadata::PermutedTriple& triple,
+                            size_t ignoreIndex = 3) {
+  const Id& undefined = Id::min();
+  switch (ignoreIndex) {
+    case 3:
+      return std::make_tuple(triple.col0Id_, triple.col1Id_, triple.col2Id_);
+    case 2:
+      return std::make_tuple(triple.col0Id_, triple.col1Id_, undefined);
+    case 1:
+      return std::make_tuple(triple.col0Id_, undefined, undefined);
+    case 0:
+      return std::make_tuple(undefined, undefined, undefined);
+    default:
+      // ignoreIndex out of bounds
+      AD_FAIL();
+  }
+};
+
+//______________________________________________________________________________
+// Check required conditions.
+static void checkEvalRequirements(const std::vector<BlockMetadata>& input,
+                                  size_t evaluationColumn) {
+  const auto throwRuntimeError = [](const std::string& errorMessage) {
+    throw std::runtime_error(errorMessage);
+  };
+  // Check for duplicates.
+  if (auto it = std::ranges::adjacent_find(input); it != input.end()) {
+    throwRuntimeError("The provided data blocks must be unique.");
+  }
+  // Helper to check for fully sorted blocks. Return `true` if `b1 < b2` is
+  // satisfied.
+  const auto checkOrder = [](const BlockMetadata& b1, const BlockMetadata& b2) {
+    return b1.blockIndex_ < b2.blockIndex_ &&
+           getMaskedTriple(b1.lastTriple_) < getMaskedTriple(b2.firstTriple_);
```

Comment:
The && check here is misleading. Just work on the block index.
If you want to you can additionally assert (via AD_C|ORRECTNE|SS_CHECK) some invarinants:

* If ` index1 < index2` then `lastTriple1` <= `firstTriple2`. Additionally the other way round (for `>` and `==`)
Because this is exactly what you are assuming (lastTriple < nextFirstTriple currently holds, but wouldn't break your code if it should ever change.